### PR TITLE
refactor: add retry for flight service

### PR DESCRIPTION
**Pull request by dqhl76**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

**Support Retry for Cluster Connection: If a network issue is detected, the system will automatically establish a new connection and attempt to continue operations.**

**Changes Details:**

- The existing implementation of the `FlightService` lacks support for encoding data with `Arc`. This leads to inefficient and resource-intensive cloning operations. Vendor the FlightService trait to develop and implement a custom codec that supports `Arc` encoding.
- Change the `do_get` method to `do_exchange` to enable a bidirectional streaming connection, allowing data to be sent from the client side to the server side.

## Tests

- [x] Unit Test
- [ ] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [x] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/databendlabs/databend/16234)

### CODE REVIEW

File: src/query/service/src/servers/flight/flight_service.rs [hunk 1]
```diff
@@ -12,96 +12,432 @@
...
+    async fn get_flight_info(
+        &self,
+        request: tonic::Request<FlightDescriptor>,
+    ) -> Result<tonic::Response<FlightInfo>, Status>;
+
+    async fn get_schema(
+        &self,
+        request: tonic::Request<FlightDescriptor>,
+    ) -> Result<tonic::Response<SchemaResult>, Status>;
+    type DoExchangeStream: tokio_stream::Stream<Item = Result<FlightData, Status>> + Send + 'static;
```
Comment (zhang2014) [inline src/query/service/src/servers/flight/flight_service.rs:78]: Arc<FlightData>?

**Comment by dqhl76** [comment]
Continue from https://github.com/datafuselabs/databend/pull/16168

File: src/query/service/src/servers/flight/flight_client.rs [hunk 2]
```diff
@@ -276,43 +463,267 @@ impl FlightSender {
...
+        if state.finish {
+            self.notify.notify_waiters();
+            state.receiver.close();
+            return;
         }
+        let weak_state = Arc::downgrade(&self.state);
+        let notify = Arc::downgrade(&self.notify);
+        let handle = GlobalIORuntime::instance().spawn(async move {
+            tokio::time::sleep(Duration::from_secs(60)).await;
+            if let Some(ss) = weak_state.upgrade() {
```
Comment (zhang2014) [inline src/query/service/src/servers/flight/flight_client.rs:711]:
```suggestion
            if let Some(state) = weak_state.upgrade() {
```

File: src/query/service/src/servers/flight/flight_service.rs [hunk 3]
```diff
@@ -1,3 +1,4 @@
+// Copyright 2016-2019 The Apache Software Foundation
```
Comment (zhang2014) [inline src/query/service/src/servers/flight/flight_service.rs:1]: How can we more friendly reference this code? cc: @Xuanwo

File: src/query/service/src/servers/flight/v1/exchange/exchange_manager.rs [hunk 4]
```diff
@@ -607,97 +713,127 @@ impl QueryCoordinator {
...
-        let (tx, rx) = async_channel::bounded(8);
-        self.fragment_exchanges.insert(
-            (target, fragment, FLIGHT_SENDER),
-            FlightExchange::create_sender(tx),
-        );
-        Ok(rx)
+        begin: usize,
+        client_stream: Streaming<FlightData>,
+    ) -> Result<FlightDataAckStream> {
+        let (tx, rx) = async_channel::unbounded();
```
Comment (zhang2014) [inline src/query/service/src/servers/flight/v1/exchange/exchange_manager.rs:719]:
```suggestion
        let (tx, rx) = async_channel:: bounded(8);
```

File: src/query/service/src/servers/flight/flight_service.rs (hunk 3 above)
Comment (Xuanwo) [reply src/query/service/src/servers/flight/flight_service.rs:1]: The current way looks good enough.

**Comment by dqhl76** [comment]
add a switch setting for retry behaviour:

```
root@localhost:8000/default> SET GLOBAL max_flight_connection_retry_times=0;

0 row read in 0.085 sec. Processed 0 row, 0 B (0 row/s, 0 B/s)
```
```
root@localhost:8000/default> select avg(number) from numbers(3000000000);
error: APIError: ResponseError with 1038: h2 protocol error: error reading a body from connection: connection reset, source: Some(hyper::Error(Body, Error { kind: Io(Kind(ConnectionReset)) }))
```

**Review by zhang2014 — APPROVED** [review]

**Pull request closed without merging**